plutus-tx/src/PlutusTx/Builtins/Class.hs

```diff
@@ -220,4 +165,80 @@ instance ToBuiltin BLS12_381.Pairing.MlResult BuiltinBLS12_381_MlResult where
     {-# INLINABLE toBuiltin #-}
     toBuiltin = BuiltinBLS12_381_MlResult
 
+{- Note [Builtin types and their Haskell versions]
+Consider the bulitin pair type. In Plutus Tx, we have an (opaque) type for
+this. It's opaque because you can't actually pattern match on it, instead you can
+only in fact use the specific functions that are available as builtins.
+
+We _also_ have the normal Haskell pair type. This is very different: you can
+pattern match on it, and you can use whatever user-defined functions you like on it.
+
+Users would really like to use the latter, and not the former. So we often want
+to _wrap_ our builtin functions with little adapters that convert between the
+"opaque builtin" "version" of a type and the "normal Haskell" "version" of a type.
 
+This is what the ToBuiltin and FromBuiltin classes do. They let us write wrappers
+for builtins relatively consistently by just calling toBuiltin on their arguments
+and fromBuiltin on the result. They shouldn't really be used otherwise.
+
+To keep the consistency we define dummy instances that are just the identity function
+for types that don't have a separate "normall Haskell" version. For example:
+integer. Integer in Plutus Tx user code _is_ the opaque builtin type, we don't
+expose a different one. Essentially: if it's a datatype then there's a substantive
+conversion.
```

Comment:
This sounds like we have two cases:

1. `Bool` vs `BuiltinBool` -- Plutus Tx type _vs_ built-in type, both appear in the `ToBuiltin`/`FromBuiltin` instance
2. `Integer` vs `BuiltinInteger` -- Plutus Tx type _and_ built-in type, both appear in the `ToBuiltin`/`FromBuiltin` instance

but we actually have a third one:

3. `Data` vs `BuiltinData` -- Plutus Tx type _vs_ built-in type, only the latter appears in the `ToBuiltin`/`FromBuiltin` instance

So this Note doesn't clarify this particular quirk for me.

@zliu41 did provide some explanations in [slack](https://input-output-rnd.slack.com/archives/G01EE4NQ9U0/p1695125397674719?thread_ts=1695059755.064939&cid=G01EE4NQ9U0), so could you please incorporate those if you agree with them?

Reply:
Hmm, I agree that the `Data` case is odd. I am actually not sure why it's like that. I'm going to try changing it and see what breaks.

Reply:
It's actually quite odd, I'm trying to pin down how it should work.